### DISCUSSION TITLE
Bundle ".wasm" in HTML file

### DISCUSSION THREAD

**Discussion by mattiaferrari02** [webpack/webpack#16360] [Q&A]
I need to bundle all my js code in a single html file.
To achieve that I'm using "html-inline-script-webpack-plugin".
It turns out that one of my deps in node_modules now uses WebAssembly and now when i run "webpack --mode production" I obtain an HTML file + a ".wasm" file.
Is there a way to pack also the wasm file in html

**Comment by alexander-akait** [comment]
Can you clarify? What do you mean pack? Can you provide configuration which you use right now?

**Reply by matthiasgeihs** [reply]
I think I have a similar issue.

By default, with webpack 5 and `asyncWebAssembly: true`, webpack seems to generate some glue code that dynamically loads the wasm on module startup.

However, this seems to assume that the wasm is served from a certain location at the server. I would like to package this in a way that doesn't require the wasm to be served separately.

For example, this would be possible (I think) through inlining the WASM. Does WebPack support something like this?

**Reply by alexander-akait** [reply]
Do you mean handle WASM file as a file and load it by self?

**Reply by matthiasgeihs** [reply]
I mean something like `asyncWebAssemblyInline: true`. And then, when writing `import factorial from "factorial.wasm";`, WebPack generates code that inlines the WASM (like with `asset/inline`), and loads the WASM module from the inlined WASM (instead of trying to fetch it from the server). Does that make sense?

**Reply by alexander-akait** [reply]
Sorry for delay, you a loader for such purposes. Why you need this, because some optimizations are impossible in the such case

**Reply by matthiasgeihs** [reply]
We have been running into problems when importing the `wasm` package into other packages with other contexts. In these cases, there might be no webserver available that servers the wasm. Nevertheless, we, ideally, wanted to have 1 bundle that we can distribute everywhere.

**Reply by matthiasgeihs** [reply]
I've been able to achieve this by packaging the WASM as `asset/inline` (as described [here](https://stackoverflow.com/questions/70420273/how-can-i-make-webpack-embed-my-wasm-for-use-in-a-web-worker)), but my solution requires some custom extra code to load the WASM.
```
import * as wasmLib from "./pkg";
import wasmDataURL from "./pkg/example.wasm";

const base64Data = (wasmDataURL as any).split("base64,")[1]; // Extract base64 string from DataURL.
const wasmBuffer = Uint8Array.from(atob(base64Data), (c) => c.charCodeAt(0)); // Convert base64 to Uint8Array.

wasmLib.initSync(wasmBuffer); // Init WASM library.
```

It would be cool if this glue code wasn't needed.

**Reply by alexander-akait** [reply]
It is very specific, I don't think we really need an extra module type for this, it is just a couple line of code, also you can write a simple custom **pitch** loader for this
